### ISSUE TITLE
Allow setting a final instance attribute in `__set_name__`?

### ISSUE THREAD

**Issue by tyralla** [python/mypy#15772]
Currently, Mypy reports the following problems:

```python
from typing import Any, Final, Type

class D:
    name: Final[str]  # error: Final name must be initialized with a value  [misc]

    def __set_name__(self, owner: Type[Any], name: str) -> None:
        self.name = name  # error: Cannot assign to final attribute "name"  [misc]
```

I know [PEP 591](https://peps.python.org/pep-0591/) does not mention `__set_name__`, but I think it could be helpful if Mypy would handle it like `__init__` regarding the one-time initialisation of instance attributes.

**Comment by sobolevn** [comment]
The problem with `__set_name__` vs `__init__` is that `__set_name__` can be skipped.

Simplest repro:

```python
reveal_type(D().name)  # Revealed type is "builtins.str"
```

This is not true: 

```python
>>> from typing import Any, Final, Type
>>> 
>>> class D:
...     name: Final[str]
...     def __set_name__(self, owner: Type[Any], name: str) -> None:
...         self.name = name
... 
>>> D().name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'D' object has no attribute 'name'
```

So, I guess some error here is better than none. In this case you can add some default like `''` or `None` here and make it mutable (because it is).

**Comment by tyralla** [comment]
Yes, you are right.  It is probably unrealistic to statically check if `__set_name__` is called for all instances of a descriptor that defines one attribute as final but initialises it in `__set_name__` instead of  `__init__`.

However, are such checks not generally outside Mypy's scope?  The following example crashes, too, without Mypy reporting any problems:

```python
class A:
    x: int

reveal_type(A().x)
```

Or is there a general promise that final attributes are always available after initialisation, which does not apply to mutable attributes?

**Comment by sobolevn** [comment]
I think that `Final` has more guarantees by design. For example, you can `del x.your_mutable_attr` and it is not a big deal. It is not the same for `Final`

**Comment by tyralla** [comment]
Deleting attributes in the context of static typing is a weird thing.  (But probably, Mypy has good reasons not to report errors for such cases).

Anyway, thanks for your thoughts on this.  If they reflect the opinion of the core team, we can readily close this issue before diving into philosophical discussions...

**Comment by sobolevn** [comment]
Let's wait for more opinions, I try to never push my views (they are very specific) :)

**Comment by huix11** [comment]
o

**Comment by huix11** [comment]
1
